### NO 5
Melakukan test hipotesis 2 sampel

In [3]:
import scipy.stats as st
import pandas as pd
import matplotlib.pyplot as plt
from statsmodels.stats.weightstats import ztest
from statsmodels.stats.proportion import proportions_ztest


In [4]:
col_names = ['id', 'pH', 'Hardness', 'Solids', 'Chloramines', 'Sulfate',
             'Conductivity', 'OrganicCarbon', 'Trihalomethanes', 'Turbidity', 'Potability']
df = pd.read_csv('../data/water_potability.csv', names=col_names)


#### a) Data kolom Sulfate dibagi 2 sama rata: bagian awal dan bagian akhir kolom. Benarkah rata-rata kedua bagian tersebut sama?

$H_0$: Nilai rata-rata kolom awal sulfate sama dengan nilai rata-rata kolom akhir sulfate $(\mu_1 - \mu_2 = 0)$

$H_1$: Nilai rata-rata kolom awal sulfate tidak sama dengan nilai rata-rata kolom akhir sulfate $(\mu_1 - \mu_2 \ne 0)$

Tingkat signifikan $\alpha = 0.05$

Lakukan pengujian two tailed test karena akan dicek pada bagian kiri dengan $z < -z_{\alpha/2}$ serta bagian kanan dengan $z > z_{\alpha/2}$  

Hitung nilai z:
$$ z = \frac{(\bar{x_1} - \bar{x_2}) - (\mu_1 - \mu_2)}{\sqrt{{\sigma_1^2/n_1} + {\sigma_2^2/n_2}}} $$

Tolak $H_0$ jika $z > z_{\alpha/2}$ atau $z < -z_{\alpha/2}$ dan $p < \alpha$  
Terima $H_0$ jika $-z_{\alpha/2} \le z \le z_{\alpha/2}$ dan $p \ge \alpha$

In [10]:
delta = 0 
alpha = 0.05

#calculate z and p using ztest module from beginning and end
len_data_per_2 = len(df) // 2
z, p = ztest(df["Sulfate"][:len_data_per_2], df["Sulfate"][len_data_per_2:], value=delta)

#calculate z_alpha/2
z_a = st.norm.ppf(1-alpha/2)

print(f"Nilai z: {round(z, 4)}")
print(f"Nilai z_alpha/2: {round(z_a, 4)}")
print(f"Nilai p: {round(p, 4)}")


Nilai z: -2.0753
Nilai z_alpha/2: 1.96
Nilai p: 0.038


Nilai $z$ lebih kecil dibandingkan $-z_{\alpha/2}$ $(-2.0753 < 1.96)$  
Nilai $p$ lebih kecil dibandingkan $\alpha$  $(0.038 < 0.05)$  
Maka tolak $H_0$

**Kesimpulan: rata-rata bagian awal dan akhir kolom sulfat ialah tidak sama**

#### b) Data kolom OrganicCarbon dibagi sama rata: bagian awal dan bagian akhir kolom. Benarkah rata-rata bagian awal lebih besar daripada bagian akhir sebesar 0.15?

$H_0$: Nilai rata-rata kolom awal organic carbon dikurang rata-rata kolom akhir organic carbon $(\mu_1 - \mu_2 = 0.15)$

$H_1$: Nilai rata-rata kolom awal organic carbon lebih dari nilai rata-rata kolom akhir organic carbon sebesar 0.15 $(\mu_1 - \mu_2 \ne 0.15)$

Tingkat signifikan $\alpha = 0.05$

Lakukan pengujian two tailed test karena akan dicek pada bagian kiri dengan $z < -z_{\alpha/2}$ serta bagian kanan dengan $z > z_{\alpha/2}$  

Hitung nilai z:
$$ z = \frac{(\bar{x_1} - \bar{x_2}) - (\mu_1 - \mu_2)}{\sqrt{{\sigma_1^2/n_1} + {\sigma_2^2/n_2}}} $$

Tolak $H_0$ jika $z > z_{\alpha/2}$ atau $z < -z_{\alpha/2}$ dan $p < \alpha$  
Terima $H_0$ jika $-z_{\alpha/2} \le z \le z_{\alpha/2}$ dan $p \ge \alpha$

In [13]:
delta = 0.15
alpha = 0.05

#calculate z and p using ztest module from beginning and end
len_data_per_2 = len(df) // 2
z, p = ztest(df["OrganicCarbon"][:len_data_per_2],
             df["OrganicCarbon"][len_data_per_2:], value=delta)

#calculate z_alpha/2
z_a = st.norm.ppf(1-alpha/2)

print(f"Nilai z: {round(z, 4)}")
print(f"Nilai z_alpha/2: {round(z_a, 4)}")
print(f"Nilai p: {round(p, 4)}")


Nilai z: -2.4131
Nilai z_alpha/2: 1.96
Nilai p: 0.0158


Nilai $z$ lebih kecil dibandingkan $-z_{\alpha/2}$ $(-2.4131 < 1.96)$  
Nilai $p$ lebih kecil dibandingkan $\alpha$  $(0.0158 < 0.05)$  
Maka tolak $H_0$

**Kesimpulan: bagian awal dan bagian akhir kolom organic carbon memiliki rata - rata bagian awal lebih besar daripada bagian akhir sebesar 0.15**

#### c) Rata-rata 100 baris pertama kolom Chloramines sama dengan 100 baris terakhirnya?

$H_0$: Nilai rata-rata 100 baris pertama kolom Chloramines sama dengan rata-rata 100 baris kolom akhir Chloramines $(\mu_1 - \mu_2 = 0)$

$H_1$: Nilai rata-rata kolom awal Chloramines lebih tidak sama dengan rata-rata 100 baris kolom akhir Chloramines $(\mu_1 - \mu_2 \ne 0)$

Tingkat signifikan $\alpha = 0.05$

Lakukan pengujian two tailed test karena akan dicek pada bagian kiri dengan $z < -z_{\alpha/2}$ serta bagian kanan dengan $z > z_{\alpha/2}$  

Hitung nilai z:
$$ z = \frac{(\bar{x_1} - \bar{x_2}) - (\mu_1 - \mu_2)}{\sqrt{{\sigma_1^2/n_1} + {\sigma_2^2/n_2}}} $$

Tolak $H_0$ jika $z > z_{\alpha/2}$ atau $z < -z_{\alpha/2}$ dan $p < \alpha$  
Terima $H_0$ jika $-z_{\alpha/2} \le z \le z_{\alpha/2}$ dan $p \ge \alpha$

In [17]:
delta = 0
alpha = 0.05

#calculate z and p using ztest module from beginning and end
z, p = ztest(df["Chloramines"].head(100),
             df["Chloramines"].tail(100), value=delta)

#calculate z_alpha/2
z_a = st.norm.ppf(1-alpha/2)

print(f"Nilai z: {round(z, 4)}")
print(f"Nilai z_alpha/2: {round(z_a, 4)}")
print(f"Nilai p: {round(p, 4)}")


Nilai z: -0.7059
Nilai z_alpha/2: 1.96
Nilai p: 0.4802


Nilai $z$ berada di rentang dibandingkan $-z_{\alpha/2} \le z \le z_{\alpha/2}$ $(-1.96 \le -0.7059 \le 1.96)$  
Nilai $p$ lebih besar dibandingkan $\alpha$  $(0.4802 > 0.05)$  
Maka terima $H_0$

**Kesimpulan: Rata-rata 100 baris pertama kolom Chloramines sama dengan 100 baris terakhirnya**

#### d) Proporsi nilai bagian awal Turbidity yang lebih dari 4 adalah lebih besar daripada proporsi nilai yang sama pada di bagian akhir Turbidity?

$H_0$: Proporsi nilai bagian awal Turbidity yang lebih dari 4 sama dengan proporsi nilai bagian akhir turbidity $(p_1 - p_2 = 0)$

$H_1$: Proporsi nilai bagian awal Turbidity yang lebih dari 4 lebih dari proporsi nilai bagian akhir $(p_1 - p_2 > 0)$

Tingkat signifikan $\alpha = 0.05$

Lakukan pengujian one tailed test karena akan dicek pada bagian kanan dengan $z > z_{\alpha}$  

Hitung nilai z:
$$ z = \frac{\hat{p_1} - \hat{p_2}}{\sqrt{\hat{p}\hat{q}({1/n_1} + {1/n_2})}} $$

dengan nilai $\hat{p}$:
$$ \hat{p} = \frac{x_1 + x_2}{n_1 + n_2}, \hat{q} = 1 - \hat{p}$$

Tolak $H_0$ jika $z > z_{\alpha}$ dan $p < \alpha$  
Terima $H_0$ jika $z \le z_{\alpha}$ dan $p \ge \alpha$

In [29]:
delta = 0
alpha = 0.05

len_data_per_2 = len(df) // 2
df_beg = df[:len_data_per_2]
df_end = df[len_data_per_2:]

#calculate z and p using ztest module from beginning and end
z, p = proportions_ztest([len(df_beg[df_beg["Turbidity"] > 4]), len(df_end[df_end["Turbidity"] > 4])], [len(df_beg), len(df_end)], value=delta, prop_var=delta)

#calculate z_alpha/2
z_a = st.norm.ppf(1-alpha)

print(f"Nilai z: {round(z, 4)}")
print(f"Nilai z_alpha: {round(z_a, 4)}")
print(f"Nilai p: {round(p, 4)}")

Nilai z: -0.1339
Nilai z_alpha: 1.6449
Nilai p: 0.8935


Nilai $z$ kurang dari $z_{\alpha}$ $(-0.1339 \le 1.6449)$  
Nilai $p$ lebih besar dibandingkan $\alpha$  $(0.8935 > 0.05)$  
Maka terima $H_0$

**Kesimpulan: Proporsi nilai bagian awal Turbidity yang lebih dari 4 sama dengan proporsi nilai bagian akhir turbidity**

#### e) Bagian awal kolom Sulfate memiliki variansi yang sama pada bagian akhirnya?

$H_0$: Variansi bagian awal kolom Sulfate memiliki nilai yang sama dengan bagian akhir kolom Sulfate $(\sigma_1^2 = \sigma_2^2)$

$H_1$: Variansi bagian awal kolom Sulfate memiliki nilai yang berbeda dengan bagian akhir kolom Sulfate $(\sigma_1^2 \ne \sigma_2^2)$

Tingkat signifikan $\alpha = 0.05$

Lakukan pengujian two tailed f test karena akan dicek pada bagian kanan dengan $f > f_{\alpha/2}(v_1,v_2)$ dan bagian kiri pada $f < {f_{1-\alpha/2}(v_1, v_2)}$

Hitung nilai f:
$$ f = \frac{s_1^2}{s_2^2} $$

dengan nilai $v_1$ dan $v_2$:
$$ v_1 = n_1 - 1, v_2 = n_2 - 2$$

Tolak $H_0$ jika $f > f_{\alpha/2}(v_1,v_2)$ atau $f < {f_{1-\alpha/2}(v_1, v_2)}$ dan $p < \alpha$  
Terima $H_0$ jika ${f_{1-\alpha/2}(v_1, v_2) \le f \le f_{\alpha/2}(v_1,v_2)}$  dan $p \ge \alpha$


In [39]:
alpha = 0.05
len_data_per_2 = len(df) // 2
df_beg = df[:len_data_per_2]
df_end = df[len_data_per_2:]
v1 = len(df_beg) - 1
v2 = len(df_end) - 1

#calculate f and p using scipy module from beginning and end
f = df_beg["Sulfate"].var() / df_end["Sulfate"].var()
f_up = st.f.ppf(1 - alpha/2, v1, v2)
f_down = st.f.ppf(alpha/2, v1, v2)

# p is the area from beginning to f
p = 1 - st.f.cdf(f, v1, v2)

print(f"Nilai f: {round(f, 4)}")
print(f"Nilai f_up: {round(f_up, 4)}")
print(f"Nilai f_down: {round(f_down, 4)}")
print(f"Nilai p: {round(p, 4)}")


Nilai f: 1.0153
Nilai f_up: 1.1318
Nilai f_down: 0.8836
Nilai p: 0.4053


Nilai $f$ berada pada range ${f_{1-\alpha/2}(v_1, v_2) \le f \le f_{\alpha/2}(v_1,v_2)}$ $(0.8836 < 1.0153 < 1.1318)$  
Nilai $p$ lebih besar dibandingkan $\alpha$  $(0.4053 > 0.05)$  
Maka terima $H_0$

**Kesimpulan: Variansi bagian awal kolom Sulfate memiliki nilai yang sama dengan bagian akhir kolom Sulfate**